[Review]: Liquidations - Add angle borrowing module liquidations (mainnet)

tutacrypto opened the pull request:
**NOTE**

1. The protocol is usually listed within 24 hours of merging the PR
2. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
3. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
4. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
5. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama): Angle 


##### Twitter Link: https://twitter.com/AngleProtocol


##### List of audit links if any: https://docs.angle.money/resources/audits


##### Website Link: https://app.angle.money/


##### Logo (High resolution, preferably in .svg and .png, for application on both white and black backgrounds. Will be shown with rounded borders):
https://raw.githubusercontent.com/AngleProtocol/angle-assets/main/0_tokens/ANGLE-token/ANGLE%20Token-light.svg

All other info can be found at: https://defillama.com/protocol/angle

tutacrypto commented:
This PR needs a review: everything looks fine when printing the `positions` output, but the test doesn't pass.

0xStrobe commented:
@tutacrypto Hi, thanks for opening this PR and giving it a stab! Some issues:

- I'm not able to see the output and I'm getting `Error: invalid contract address or ENS name (argument="addressOrName", value=undefined, code=INVALID_ARGUMENT, version=contracts/5.7.0)` which is likely because your code is trying to feed `undefined` into some contract calls which should have been an address. You mentioned it looks fine printing the positions out - how to reproduce your result?
- Could you please revert the formatting in `liquity/index.ts` and `euler/index.ts`, as well as the `package.json` and the package lock file inside the root `/liquidations/` and the package lock on the repo root? I see you created a package.json and installed ts-node within `/liquidations/`, which is not needed.

I will write a short readme on adding liquidations adapters, but for now - you can use the following command to run the test script to check if your output is correct. Assuming you run this command at repo root:

```bash
npx ts-node ./liquidations/test.ts ./liquidations/liquity/index.ts
```

tutacrypto commented:
Hello @0xStrobe , thanks for the quick feedback! 

I reverted `liquity/index.ts` and `euler/index.ts `, and removed `package-lock.json` and `package.json` from the liquidations folder. 

I get the output by running this: (commented at the end of angle/index.ts) 
```
positions().then(vaultData => {
    console.log(vaultData);
});
```

and get this kind of output: 
```
 ...
   },
{
    owner: '0x9c0d1f4a029c46265831d120dee9cdc72f0ab3c3',
    liquidationPrice: 0.8915938376458833,
    collateral: '0x5f98805a4e8be255a32880fdec7f6728c6568ba0',
    collateralAmount: '13000000000000000000000',
    extra: {
      url: 'https://etherscan.io/address/0x9c0d1f4a029c46265831d120dee9cdc72f0ab3c3'
    }
  }
]
```

I'll investigate a bit more to see if I can make the test pass!

tutacrypto commented:
Test passes! 

Apparently the issue was that I wasn't specifying the chain before the collateral address line 86.

Let me know if there is anything else.

0xStrobe commented on liquidations/angle/index.ts:
```diff
@@ -0,0 +1,102 @@
+    const collateralFactor = parseFloat(vault.vaultManager.collateralFactor) / 10e8
+    let liquidationPrice: number
+    if (collateral == '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599') {
+      liquidationPrice = parseFloat(vaultDebt) / (parseFloat(collateralAmount)*10e9*collateralFactor)
```
Is WBTC the only token which needed converting decimals with `*10e9` and why `*10e9`? If there could be more assets which needs conversion, we prefer to read the decimals from on-chain:

```ts
const { output: decimals } = await sdk.api.erc20.decimals(token, chain);
```
... or use the results from llama token price API: https://github.com/DefiLlama/DefiLlama-Adapters/blob/de5be4926982d43e875d164e7f56676d0a644822/liquidations/utils/compound-helpers.ts#L41

0xStrobe replied:
Or you can use this piece of code to get the price from DefiLlama Token Price API

```ts
type Price = {
  decimals: number;
  symbol: string;
  price: number;
  timestamp: number;
  confidence: number;
};

// this code will return a dictionary shaped as the Price type
const getAgeurPriceInUsd = async () => {
  const tokenId = "ethereum:0x1a7e4e63778b4f12a199c062f3efdd288afcbce8";
  const res = (
    await axios.get("https://coins.llama.fi/prices/current/" + tokenId)
  ).data.coins[tokenId] as Price;
  return res;
};

```

0xStrobe commented on liquidations/angle/index.ts:
```diff
@@ -0,0 +1,102 @@
+    return vaultData;
+};
+
+const getPrices = async () => {
```
We prefer to use llama price api - https://github.com/DefiLlama/DefiLlama-Adapters/blob/de5be4926982d43e875d164e7f56676d0a644822/liquidations/utils/compound-helpers.ts#L41

0xStrobe reviewed (changes requested):
Added a couple of comments :)

I also noticed that the liquidation price returned by the adapter is always lower than the price listed on Angle's web app. Is the method used for calculating the liquidation price inaccurate?

For example:

**User 0x3f7c10cbbb1ea1046a80b738b9eaf3217410c7f6**

Adapter:
```jsonc
  {
    owner: '0x3f7c10cbbb1ea1046a80b738b9eaf3217410c7f6',
    liquidationPrice: 722.2249926624788,
    collateral: 'ethereum:0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0',
    collateralAmount: '47887301493786015873',
    extra: {
      url: 'https://etherscan.io/address/0x3f7c10cbbb1ea1046a80b738b9eaf3217410c7f6'
    }
  }
```

Webapp:
<img width="806" alt="image" src="https://user-images.githubusercontent.com/107649472/193051389-89eb7bfb-4301-4030-8487-322c05f43614.png">

tutacrypto commented:
The difference in liquidation price is because it is labelled in EUR on our frontend (as vaults debt is in agEUR), and the output in this PR is in USD.

I can't manage to access the price or decimals that I get from DeFillama API. The following code returns me `TypeError: Cannot read properties of undefined (reading 'price')`

```
const getPrices = async () => {
  //const url: string = 'https://api.coingecko.com/api/v3/simple/price?ids=ageur&vs_currencies=usd';
  const info = (
    await axios.post("https://coins.llama.fi/prices", {
      coins: ["ethereum:0x1a7e4e63778B4f12a199C062f3eFdD288afCBce8"],
    })
  ).data.coins as {
    [address: string]: { decimals: number; price: number; symbol: string; timestamp: number };
  };
  const price = info["ethereum:0x1a7e4e63778B4f12a199C062f3eFdD288afCBce8"].price
  return price
} 
```

tutacrypto commented:
Thank you for the review! 

I'm now using DefiLlama API, and I fixed the PR to use something generic for the decimal changes.

0xStrobe commented:
Thanks for the contribution! I made some improvements and merged.

0xStrobe commented:
@tutacrypto I've noticed that Angle on mainnet only have ~8 positions - would you like to add support for other chains as well, since Angle is also available on Arbitrum, Polygon and Optimism?

You can refer to our Aave V2 adapter for multichain support - https://github.com/DefiLlama/DefiLlama-Adapters/blob/cd2fad22a361565d9cf6c13899d23031bf77b59f/liquidations/aave-v2/index.ts#L141

tutacrypto commented:
It looks like some of your fixes broke the code, the test now returns an error for `angle/index.ts` 
Also I don't see neither Angle, Aave, or Benqi on the front, so I tried the tests for `Aave` and `Benqi` and they appear to be stuck (or maybe just taking a long time). 

About liquidations on multiple chains, I wanted to do mainnet correctly first and than I forgot 😅  
I'll work on it next week once we've fixed this!

0xStrobe commented:
Thanks for noticing, I indeed missed something, it's now fixed. :)

> Also I don't see neither Angle, Aave, or Benqi on the front

What do you mean? You don't see them on the frontend or the tests are not working? They are working on my side. Angle is not yet listed on the frontend, I'll enable it soon.

Yes, it's known that the liquidations adapters for big protocols with a lot of events can take a long time to run through since we process all open positions on a protocol, which can be tens of thousands on Aave for example.

tutacrypto commented:
Everything's working now, thanks!

I'll open another PR next week for the other chains and tag you when it's done :)
